pymeasure/instruments/hp/hp3478A.py

```diff
@@ -0,0 +1,316 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+
+import logging
+from pymeasure.instruments import Instrument
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+
+class HP3478A(Instrument):
+    """ Represents the Hewlett Packard 3748A 5 1/2 digit multimeter
+    and provides a high-level interface for interacting
+    with the instrument.
+
+    As this unit predates SCPI some tricks are required to get this working
+    """
+    MODES={"DCV": "F1",
+           "ACV": "F2",
+           "R2W": "F3",
+           "R4W": "F4",
+           "DCI": "F5",
+           "ACI": "F6",
+           "Rext": "F7",
+           }
+
+    RANGES={"DCV": {3E-2: "R-2", 3E-1: "R-1",3: "R0",30: "R1",300:"R2",
+                    "auto": "RA"},
+            "ACV": {3E-1: "R-1", 3: "R0",30: "R1" ,300: "R2","auto": "RA"},
+            "R2W": {30: "R1", 300: "R2", 3E3: "R3",3E4: "R4",3E5: "R5",
+                    3E6: "R6", 3E7: "R7","auto": "RA"},
+            "R4W": {30: "R1",300: "R2", 3E3: "R3",3E4: "R4",3E5: "R5",
+                    3E6: "R6", 3E7: "R7","auto": "RA"},
+            "DCI": {3E-1: "R-1",3: "R0","auto": "RA"},
+            "ACI": {3E-1: "R-1",3: "R0","auto": "RA"},
+            "Rext": {3E6: "R7","auto": "RA"},
+            }
+
+    TRIGGERS={
+        "auto": "T1",
+        "internal": "T1",
+        "external": "T2",
+        "single": "T3",
+        "hold": "T4",
+        "fast": "T5",
+        }
+
+
+    measurement_string=''
+
+    def __init__(self, resourceName, **kwargs):
+        super(HP3478A, self).__init__(
+            resourceName,
+            "Hewlett-Packard HP3478A",
+            includeSCPI = False,
+            send_end = True,
+            read_termination = "\r\n",
+            **kwargs
+        )
+
+    check_errors = Instrument.measurement(
+        "E",
+        """Checks the error status register""",
+        get_process=(lambda x: int(x)),
+        )
+
+    status = Instrument.measurement(
+        "B",
+        """Checks the status registers""",
+        get_process=(lambda x: bytes(x.encode(encoding="ASCII"))),
+        )
+
```

Comment:
`encode` returns `bytes` , why `bytes` is needed ?

In a wider review context, I would like to propose to access the status bits returned by command "B" in a more structured way. In this mode, your code is probably slower, but much more readable, in my opinion.
My proposal is to define a static method which decodes the bytes returned by "B" command through the use of ctypes and bitfield (see attached examples [test.zip](https://github.com/pymeasure/pymeasure/files/6411976/test.zip))

Let's call the static method `get_status(status, field=None)` with the following behaviour:
- `get_status(status, None)` return all the fields
- `get_status(status, "auto_zero")` return "auto_zero" field

Advantages:
- code is more readable
- `__str__` call can replace `status_readable` utility
- `status get_process` is defined as follow:` get_process=(lambda x: self.get_status(x.encode(encoding="ASCII")))`
- `auto_zero` can be rewritten as follow:
```
auto_zero = Instrument.control(
        "B",
        "Z%d",
        """ Return autozero settings on the HP3478, this property can be set """,
        get_process=(lambda x: self.get_status(x.encode(encoding="ASCII"), "auto_zero")),
        validator=strict_discrete_set,
        values=(0,1),
        )
```
- similar approach also for other property and and methods in the class, when possible
- Using `map_value` and dictionaries allow to use symbolic names as well (see `https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html`), so probably your dictionaries `MODES`, `RANGES`, etc. can be reused.
- Align code to class `Instrument` style and make it probably more homogenous to the rest of instruments.

Sorry, probably a lot of changes, but if you think are useful, it could be worth to make them.

Reply:
@msmttchr  Thank you again for the code review, I really appreciate your support in this matter.

> encode returns bytes , why bytes is needed ?


The 'B' command returns 5 bytes which are not ment to be read as an ASCII encoding.
The Intstrument.control()/messurement()/setting() functions however do not offer a passing of binary values like Instrument.binary_read() does, this is why I used the encode call.
For reference I attach a exerpt of the manual 
![3478A_manual_pg_3-14](https://user-images.githubusercontent.com/17066917/116834062-6d107900-abf7-11eb-8358-2f00b69a2777.png)

> In a wider review context, I would like to propose to access the status bits returned by command "B" in a more structured way. In this mode, your code is probably slower, but much more readable, in my opinion.
My proposal is to define a static method which decodes the bytes returned by "B" command through the use of ctypes and bitfield (see attached examples test.zip)

I like your idea of a static method to decode the bits from the status bytes.
Speed is not an issue here, as the status of the instrument is normally not queried very quickly.
And btw this instrument is anyway not the fastest (being from the early '80ties)
Where speed was concerned was the measure routine, which now uses caching of the command to be sent.

Looking at your proposal with the ctypes, I have been looking at enum over the last days and am working on that route at the moment, I have attached an example 
[enum_test.zip](https://github.com/pymeasure/pymeasure/files/6412715/enum_test.zip)
(The function and range encoding still gives me some headache, as the bit-assignment is not unique)


**General question:** is it save to assume that a python version >3.6 will be used? Or is there is a feature set we should respect.
For my private use I am at Pyhton 3.9.4 (ArchLinux/ArchLinuxArm) and for my workstuff typically also on >3.8(WInpython), but I dont know if this can be expected for all pymeasure users.

Reply:
@LongnoseRob, possibly I was too crytpic in my comment, here I meant that the code below is equivalent:
- `get_process=(lambda x: bytes(x.encode(encoding="ASCII"))),`
- `get_process=(lambda x: x.encode(encoding="ASCII")),`

In other word, I am not sure if the `bytes` instantiation is redundant since after call to method `encode` you have already `bytes` class data.

Regarding whether it is safe to assume on python version >3.6, from what I see in the CI tests that are run after a commit, the python version used are: 3.6, 3.7, 3.8, 3.9. In addition to this, file [setup.py](https://github.com/pymeasure/pymeasure/blob/master/setup.py) mention the same python versions.

I would assume then that python version >= 3.6.